src/Reopt/TypeInference/Header.hs

```diff
@@ -145,15 +145,34 @@ parseStructUnion (C.CStruct tag _mi _mdecl _attrs n) =
     C.CStructTag -> errorAt n "Struct is not supported."
     C.CUnionTag -> errorAt n "Union is not supported."
 
--- | Parser derived declarators.
+parseTypeDecl :: C.NodeInfo -> C.CDeclarationSpecifier C.NodeInfo -> CParser C.CTypeSpec
+parseTypeDecl _ (C.CTypeSpec spec) = return spec
+parseTypeDecl n _ = errorAt n "Expected type specification"
+
+parseTypeDecls :: C.CDeclaration C.NodeInfo -> CParser C.CTypeSpec
+parseTypeDecls (C.CDecl [decl] _ n) = parseTypeDecl n decl
+-- NOTE (val) So far I'm only seeing singleton lists here, not sure why.
+parseTypeDecls (C.CDecl _decls _ n) = errorAt n "Expected single type spec, investigate."
+parseTypeDecls (C.CStaticAssert _ _ n) = errorAt n "TODO"
+
+-- | Parse derived declarators.
 parseTypeDerivedDecl :: [C.CDerivedDeclarator C.NodeInfo] -> AnnType -> CParser AnnType
 parseTypeDerivedDecl [] tp = pure tp
 parseTypeDerivedDecl (C.CPtrDeclr _ _ : rest) tp = do
   parseTypeDerivedDecl rest $! PtrAnnType tp
 parseTypeDerivedDecl (C.CArrDeclr _ _ n : _) _tp = do
   errorAt n "Arrays are not supported."
-parseTypeDerivedDecl (C.CFunDeclr _ _ n : _) _tp = do
-  errorAt n "Function declarations are not supported in this context."
+parseTypeDerivedDecl (C.CFunDeclr (Right (decls, False)) _attrs _n : rest) tp = do
+  typeSpecs <- mapM parseTypeDecls decls
+  args <- mapM (parseType emptyQualMods) typeSpecs
+  parseTypeDerivedDecl rest $! FunPtrAnnType tp args
+  -- errorAt n $ "decls: " ++ show decls ++ "\nattrs: " ++ show attrs ++ "\ntp: " ++ show tp
+parseTypeDerivedDecl (C.CFunDeclr (Right (_, True)) _boop n : _) _tp = do
+  errorAt n "True"
+parseTypeDerivedDecl (C.CFunDeclr (Left {}) _boop n : _) _tp = do
+  errorAt n "Left"
+-- parseTypeDerivedDecl (C.CFunDeclr _ _ n : _) _tp = do
+--   errorAt n "Function declarations are not supported in this context."
```

Comment:
The commented-out code seems to give a more human-friendly error message... should that be used instead?

Reply:
Oh yeah this was just for debugging purposes, removing!